image/oci_test.go

```diff
@@ -0,0 +1,353 @@
+			f: func(digest string) (io.ReadCloser, int64, error) {
+				return ioutil.NopCloser(bytes.NewReader(realConfigJSON)), int64(len(realConfigJSON)), nil
+			},
+		},
+		ref: ref,
+	}
+}
+
+func TestManifestOCI1UpdatedImage(t *testing.T) {
+	originalSrc := newOCI1ImageSource(t, "httpd:latest")
+	original := manifestOCI1FromFixture(t, originalSrc, "oci1.json")
+
+	// LayerInfos:
+	layerInfos := append(original.LayerInfos()[1:], original.LayerInfos()[0])
+	res, err := original.UpdatedImage(types.ManifestUpdateOptions{
+		LayerInfos: layerInfos,
+	})
+	require.NoError(t, err)
+	assert.Equal(t, layerInfos, res.LayerInfos())
+	_, err = original.UpdatedImage(types.ManifestUpdateOptions{
+		LayerInfos: append(layerInfos, layerInfos[0]),
+	})
+	assert.Error(t, err)
+
+	// ManifestMIMEType:
+	// Only smoke-test the valid conversions, detailed tests are below. (This also verifies that “original” is not affected.)
+	for _, mime := range []string{
+		manifest.DockerV2Schema2MediaType,
+	} {
+		_, err = original.UpdatedImage(types.ManifestUpdateOptions{
+			ManifestMIMEType: mime,
+			InformationOnly: types.ManifestUpdateInformation{
+				Destination: &memoryImageDest{ref: originalSrc.ref},
+			},
+		})
+		assert.NoError(t, err, mime)
+	}
+	for _, mime := range []string{
+		imgspecv1.MediaTypeImageManifest, // This indicates a confused caller, not a no-op.
+		"this is invalid",
+	} {
+		_, err = original.UpdatedImage(types.ManifestUpdateOptions{
+			ManifestMIMEType: mime,
+		})
+		assert.Error(t, err, mime)
+	}
+
+	// m hasn’t been changed:
+	m2 := manifestOCI1FromFixture(t, originalSrc, "oci1.json")
+	typedOriginal, ok := original.(*manifestOCI1)
+	require.True(t, ok)
+	typedM2, ok := m2.(*manifestOCI1)
+	require.True(t, ok)
+	assert.Equal(t, *typedM2, *typedOriginal)
+}
+
+func TestConvertToManifestSchema2(t *testing.T) {
+	originalSrc := newOCI1ImageSource(t, "httpd-copy:latest")
+	original := manifestOCI1FromFixture(t, originalSrc, "oci1.json")
+	memoryDest := &memoryImageDest{ref: originalSrc.ref}
+	res, err := original.UpdatedImage(types.ManifestUpdateOptions{
+		ManifestMIMEType: manifest.DockerV2Schema2MediaType,
+		InformationOnly: types.ManifestUpdateInformation{
+			Destination: memoryDest,
+		},
+	})
+	require.NoError(t, err)
+
+	convertedJSON, mt, err := res.Manifest()
+	require.NoError(t, err)
+	assert.Equal(t, manifest.DockerV2Schema2MediaType, mt)
+
+	// byDockerJSON is the result of asking the Docker Hub for a schema1 manifest,
+	// except that we have replaced "name" to verify that the ref from
+	// memoryDest, not from originalSrc, is used.
+	byDockerJSON, err := ioutil.ReadFile("fixtures/oci1-to-schema2.json")
+	require.NoError(t, err)
+	var converted, byDocker map[string]interface{}
+	err = json.Unmarshal(byDockerJSON, &byDocker)
+	require.NoError(t, err)
+	err = json.Unmarshal(convertedJSON, &converted)
+	require.NoError(t, err)
+	delete(byDocker, "signatures")
+	delete(converted, "signatures")
+	assert.Equal(t, byDocker, converted)
```

Comment:
There should be no `signatures` fields in these versions, that’s a Docker schema1 thing.